### PR TITLE
Join space with invitation code

### PR DESCRIPTION
**Pull request by cp-ishita-g**
- [x] design join space code fields
- [x] api call for join space 
- [x] api call to get invitation code of selected space
- [x] manage last selected space when we open space next time
- [x] streaming live data for member joining

invite code
join space

### CODE REVIEW

File: app/assets/locales/app_en.arb [hunk 1]
```diff
@@ -67,6 +67,8 @@
   "join_space_title": "Join space",
   "join_space_invite_code_title": "Enter the invite code",
   "join_space_get_code_from_space_text": "Get the code from the space creator to join.",
+  "join_space_already_joined_error_text": "You are already member of that space",
```
Comment (cp-radhika-s) [inline app/assets/locales/app_en.arb:70]: You are already member of this space

File: app/lib/ui/flow/home/components/home_top_bar.dart [hunk 2]
```diff
@@ -12,17 +12,21 @@ import 'package:yourspace_flutter/ui/app_route.dart';
 import '../../../../gen/assets.gen.dart';
 
 class HomeTopBar extends StatefulWidget {
-  final void Function(String) onSpaceItemTap;
+  final void Function(SpaceInfo) onSpaceItemTap;
+  final void Function() onAddMemberTap;
   final List<SpaceInfo> spaces;
   final String title;
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/components/home_top_bar.dart:18]: We need to pass the selected `space Id` as well.
From space Id we can get details of current space so we don't need `title` param

File: app/lib/ui/flow/home/components/home_top_bar.dart [hunk 3]
```diff
@@ -12,17 +12,21 @@ import 'package:yourspace_flutter/ui/app_route.dart';
...
+  final bool isCodeGetting;
 
   const HomeTopBar({
     super.key,
     required this.spaces,
     required this.onSpaceItemTap,
+    required this.onAddMemberTap,
     required this.title,
     this.loading = false,
+    this.isCodeGetting = false,
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/components/home_top_bar.dart:29]: isCodeGetting -> fetchingInviteCode

File: app/lib/ui/flow/home/home_screen_viewmodel.dart [hunk 4]
```diff
@@ -10,42 +11,62 @@ final homeViewStateProvider = StateNotifierProvider.autoDispose<
...
+    } catch (error, stack) {
+      logger.e(
+        'HomeViewNotifier: Error while getting invitation code',
+        error: error,
+        stackTrace: stack,
       );
     }
   }
+
+  void updateSelectedSpaceName(SpaceInfo space) {
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/home_screen_viewmodel.dart:64]: updateSelectedSpaceName -> updateSelectedSpace

File: app/lib/ui/flow/space/join/join_space_view_model.dart [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    try {
+      state = state.copyWith(verifying: true);
+      final invitation = await spaceInvitationService.getInvitation(code);
+      if (invitation == null) {
+        state =
+            state.copyWith(errorInvalidInvitationCode: true, verifying: false);
+        return;
+      }
+      var spaceId = invitation.space_id;
+      final joinSpaces = authService.currentUser?.space_ids ?? [];
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/space/join/join_space_view_model.dart:40]: joinSpaces -> userSpaces

File: data/lib/service/space_service.dart [hunk 6]
```diff
@@ -69,33 +75,41 @@ class SpaceService {
...
 
-  Future<SpaceInfo?> getCurrentSpaceInfo() async {
-    final currentSpace = await getCurrentSpace();
-    if (currentSpace == null) return null;
-    final members = await spaceService.getMembersBySpaceId(currentSpace.id);
-    return SpaceInfo(
-      space: currentSpace,
-      members: members
-          .map((member) async {
+  Stream<SpaceInfo?> streamCurrentSpaceInfo() async* {
```
Comment (cp-radhika-s) [inline data/lib/service/space_service.dart:88]:
```
 Stream<List<ApiSpace>> userSpacesStream(String userId) {
    return _apiSpaceService
        .getSpaceMemberByUserId(userId)
        .asyncMap(((List<ApiSpaceMember> members) async {
      print("XXX members ${members.length}");
      var spaces = await Future.wait(members
          .map((member) async => _apiSpaceService.getSpace(member.space_id)));
      return spaces.whereType<ApiSpace>().toList();
    }));
  }
  
 Future<Stream<List<SpaceInfo>>> getAllSpaceInfo() async {
    final String userId = _authRepository.currentUser?.id ?? '';
    print("XXX userId ${userId}");
    return userSpacesStream(userId).asyncMap((List<ApiSpace> spaces) async {
      var spaceInfos = await Future.wait(spaces.map((space) async {
        List<ApiSpaceMember> members =
            await _apiSpaceService.spaceMemberStream(space.id).first;
        var userInfos = await Future.wait(members.map((member) async {
          var user = await _apiUserService.getUser(member.user_id);
          return user == null
              ? null
              : UserInfo(
                  user: user, isLocationEnabled: member.location_enabled);
        }).toList());

        userInfos.removeWhere((userInfo) => userInfo == null);

        return SpaceInfo(apiSpace: space, apiUser: userInfos);
      }).toList());

      print("XXX getAllSpaceInfo called ${spaceInfos.length}");
      return spaceInfos;
    });
  }
```

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

File: data/lib/service/space_service.dart (hunk 6 above)
Comment (cp-ishita-g) [reply data/lib/service/space_service.dart:88]: i don't think so we need to make stream func here also, by streaming get member details calls we are getting live updates

File: app/lib/ui/flow/home/components/home_top_bar.dart (hunk 2 above)
Comment (cp-ishita-g) [reply app/lib/ui/flow/home/components/home_top_bar.dart:18]: we have state of selected space from there we can get space id to get current space data

File: app/lib/ui/flow/space/join/join_space_view_model.dart [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+      if (userSpaces.contains(spaceId)) {
+        state = state.copyWith(verifying: false, alreadySpaceMember: true);
+        return;
+      }
+
+      spaceService.joinSpace(spaceId);
+      state = state.copyWith(verifying: false, pop: true);
+    } catch (error, stack) {
+      logger.e(
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/space/join/join_space_view_model.dart:50]: Notify the user with an appropriate message on failure, do the changes wherever require

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

File: app/lib/ui/flow/home/components/home_top_bar.dart [hunk 8]
```diff
@@ -52,7 +52,7 @@ class _HomeTopBarState extends State<HomeTopBar> with TickerProviderStateMixin {
       parent: _animationController,
       curve: Curves.easeInOut,
     );
-    selectedIndex = _getSelectedIndex();
+    selectedIndex = _getSelectedSpace();
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/components/home_top_bar.dart:55]: We need to show the selected space always on top, so I think we don't need `selectedIndex`

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

**Comment by cp-radhika-s** [comment]
We need to show a success popup on Space joined. Check Android flow

File: app/lib/ui/flow/home/components/home_top_bar.dart [hunk 9]
```diff
@@ -56,6 +60,11 @@ class _HomeTopBarState extends State<HomeTopBar> with TickerProviderStateMixin {
     super.dispose();
   }
 
+  @override
+  void didUpdateWidget(HomeTopBar oldWidget) {
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/components/home_top_bar.dart:64]: Do we need this?

File: app/lib/ui/flow/space/join/join_space_screen.dart [hunk 10]
```diff
@@ -63,43 +87,126 @@ class _JoinSpaceState extends ConsumerState<JoinSpace> {
...
+        ),
+      ),
+    );
+  }
+
+  Widget _joinSpaceButton(BuildContext context, JoinSpaceViewState state) {
+    return BottomStickyOverlay(
+      child: Column(
+        children: [
+          _joinSpaceError(context, state),
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/space/join/join_space_screen.dart:154]: The error shouldn't move UI, Show an error as a popup like a banner. Check android implementation

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

**Review by cp-radhika-s — APPROVED** [review]

**Pull request merged**